pagefaultgames/pokerogue
# [Qol][Refactor] i18n lazy-loading

> [!CAUTION]
> ### Make sure to consult the *@pagefaultgames/translation-team* & *@pagefaultgames/translation-leads*  before merging as all locale files moved from `/src/locales` into `/public/locales` !!!

> [!TIP]
> there's only `~7` relevant files, the rest are all file moves/etc
> - [@types/i18next.d.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-d78a60876ad1873a6f1f26a69962182be27b51f07bb9606bb6209e6912ebf319)
> - [loading-scene.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-99d6ceeb58f5b1d3a15a324e96b2a7311ef041cab023c56107c3e61badc4f8ea)
> - [main.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3d)
> - [plugins/i18n.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-60258ac1f55717ab75876c0d8a946e2880a81c964b78b44ba15aa269861de544)
> - [test/utils/mocks/mocksContainer/mockText.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-245f89be077f680057ab52be026cbcd710fc2f2b5f1eb099a2d8bbfba31337bd)
> - [test/vitest.setup.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-7ccd55f6c0451a683566386d23e2ea3afbbd3fc41edb82816520318dc4d16740)
> - [utils.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351)
>
>  \- *NightKev*

---

## What are the changes the user will see?

Only the languages being used will be loaded thus less memory usage from the i18n side 

## Why am I making these changes?

The memory shouldn't be bloated with all languages at all time. IT should only hold the languages currently being used which at max is 2 (en as a fallback for missing keys).

## What are the changes from a developer perspective?

1. Moved all `/src/locales` into `/public/locales` as they will be loaded async (via network request).
2. Install the [i18next-http-backend](https://github.com/i18next/i18next-http-backend) plugin which is necessary for lazy-loading
3. Add a `camelCaseToKebabCase` function to utils
4. Clean up `/src/plugin/i18n.ts` file structure
5. move phaser config into `startGame()` and lazy-load the scenes to prevent any `i18n.t()` usages before it is initialized
6. Remove `initI18n()` from `LoadingScene`
7. Remove all `config.ts` files inside the `locales/` 
8. **Remove all `locales` tests**. _Reason: We don't need to test if i18next is working properly. They have their own test-coverage to ensure that_
9. **`i18next` is now mocked for tests**. <sub>This was overdue and now necessary due to the backend module</sub>

## How to test the changes?

1. Play the game
2. Switch languages and keep on playing.
3. Check the network tab for any failed translation requests (`locales/**`)

## Checklist
